### PR TITLE
feat: Implement latest blocks page

### LINKED ISSUE #19: Latest blocks page
# Latest blocks page

## ⚠️ Reading contributors guidelines to get assigned is MANDATORY!

[Read contributors guidelines](https://github.com/walnuthq/op-scan/issues/1)

## User stories

- As a user, I want to see the list of the latest block created on the blockchain by visiting `/blocks`
- I want to see a table of the list of newly created blocks as well as the entire history through pagination. The detail for each block should be block number, timestamp, txns count, gas used and gas limit.

## Validation

It should look like the Etherscan latest blocks list: https://optimistic.etherscan.io/blocks

Image

For the gas used column, you should add the percentage against gas limit and you can show a progress bar using https://ui.shadcn.com/docs/components/progress
Don't forget to add links to block details page `/block/:number` and block txs `/block/:number/txs` page.

Do NOT include block metrics such as Network utilization, block size, etc.
Do NOT add a "Download Page Data" button.
Do NOT add a "Show rows" dropdown, this is out of scope for this issue.

## Implementation

Use a simple table from shadcn/ui, always display 25 blocks and use a query parameter to handle pagination:
If current block head is 100 then
- `/blocks` => blocks 100 => 76
- `/blocks?p=2` => blocks 75 => 51
- `/blocks?p=3` => blocks 50 => 26
Use client state to keep track of the latest block number upon first visiting the page, and use this as a reference for navigation: when we click on page 2, we want to see the next 25 blocks from the initial block head (100), not the next 25 blocks from the new current block head.

**You must use NextJS best practices about handling query parameters, in particular, read `searchParams` from the top-level Page component and use a React Server Component to fetch the blocks needed. Do NOT fetch the blocks client-side.**

## Resources

- shadcn/ui Table component: https://ui.shadcn.com/docs/components/table
- shadcn/ui Progress component: https://ui.shadcn.com/docs/components/progress
- NextJS docs about query parameters: https://nextjs.org/docs/app/api-reference/functions/use-search-params#server-components

### PR DESCRIPTION
**Pull request by Benjtalkshow**
closes #19   

#  Implemented paginated list of latest blocks with detailed information

## Features Implemented

- Display a table of the latest blocks using the shadcn/ui Table component.
- Each row in the table shows:
  - Block number (with a link to the block details page).
  - Timestamp (with tooltip showing the exact UTC time).
  - Transaction count (with a link to the block transactions page).
  - Gas used (with percentage against the gas limit).
  - Gas limit.
- Added a progress bar for gas used percentage using the shadcn/ui Progress component.
- Implemented pagination to navigate through the block history, displaying 25 blocks per page.
- The latest block number is fetched on the first page load and used as a reference for subsequent page navigation to ensure consistent pagination.

## Pagination Details

- Navigating to `/blocks` displays blocks 100 to 76.
- Navigating to `/blocks?p=2` displays blocks 75 to 51.
- Navigating to `/blocks?p=3` displays blocks 50 to 26.
- The latest block number is fetched using `viem getBlockNumber` if not provided as a query parameter.

## Implementation Details

- Used NextJS best practices for handling query parameters and data fetching.
- Server-side data fetching using React Server Components and `getServerSideProps`.
- Pagination is handled server-side using the `searchParams` parameter

## Screenshot Desktop
image

## Screenshot Mobile
image

Please review the changes and provide feedback. Thank you .

### CODE REVIEW

File: src/components/pages/blocks/latest-blocks-table.tsx [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+  latestBlockNumber: bigint;
+  currentPage: number;
+  blocksPerPage: number;
+}
+
+async function fetchBlocks(latestBlockNumber: bigint, currentPage: number, blocksPerPage: number) {
+  const startBlock = latestBlockNumber - BigInt((currentPage - 1) * blocksPerPage);
+  const blocks = await Promise.all(
+    Array.from({ length: blocksPerPage }, (_, i) =>
+      l2PublicClient.getBlock({ blockNumber: startBlock - BigInt(i) })
```
Comment (saimeunt) [inline src/components/pages/blocks/latest-blocks-table.tsx:33]: There's an edge case here when we're hitting the last page: if there's less than 25 blocks left to display, this code tries to get blocks with an invalid negative block number.
Make sure to guard against this edge case and display only the < 25 remaining blocks.

**Review by saimeunt — CHANGES_REQUESTED** [review]
@Benjtalkshow Thank you for your contribution, it provides a solid foundation for the latest blocks page.
There's only one change that needs to be fixed regarding an edge case for the last page of results.

File: src/components/pages/blocks/latest-blocks-table.tsx (hunk 1 above)
Comment (Benjtalkshow) [reply src/components/pages/blocks/latest-blocks-table.tsx:33]: Alright @saimeunt . Will do that ASAP. Thanks for your kind review.

**Comment by Benjtalkshow** [comment]
closes #32 
@saimeunt , I have made the requested changes to handle the edge case for fetching blocks on the last page. Please review the latest commit. Thank you!

**Review by saimeunt — APPROVED** [review]

**Pull request merged**